spec/app/ics-020-fungible-token-transfer/README.md

```diff
@@ -210,33 +253,47 @@ In plain English, between chains `A` and `B`:
   to take appropriate action based on the nature of the failure.
 
+Note: `constructOnChainDenom` is a helper function that will construct the local on-chain denomination for the bridged token. It **must** encode the trace and base denomination to ensure that tokens coming over different paths are not treated as fungible. The original trace and denomination must be retrievable by the state machine so that they can be passed in their original forms when constructing a new IBC path for the bridged token. The ibc-go implementation handles this by creating a local denomination: `hash(trace+base_denom)`.
+
 `sendFungibleTokens` must be called by a transaction handler in the module which performs appropriate signature checks, specific to the account owner on the host state machine.
 
 ```typescript
 function sendFungibleTokens(
-  denomination: string,
-  amount: uint256,
+  tokens: []Token,
   sender: string,
   receiver: string,
+  memo: string,
   sourcePort: string,
   sourceChannel: string,
   timeoutHeight: Height,
   timeoutTimestamp: uint64): uint64 {
-    prefix = "{sourcePort}/{sourceChannel}/"
-    // we are the source if the denomination is not prefixed
-    source = denomination.slice(0, len(prefix)) !== prefix
-    if source {
-      // determine escrow account
-      escrowAccount = channelEscrowAddresses[sourceChannel]
-      // escrow source tokens (assumed to fail if balance insufficient)
-      bank.TransferCoins(sender, escrowAccount, denomination, amount)
-    } else {
-      // receiver is source chain, burn vouchers
-      bank.BurnCoins(sender, denomination, amount)
+    for token in tokens {
+      prefix = "{sourcePort}/{sourceChannel}/"
+      // we are the source if the denomination is not prefixed
+      source = token.trace[0] !== prefix
+      onChainDenom = constructOnChainDenom(token.trace, token.denomination)
+      // check if we've already sent metadata for this token on this channel
+      isMetadataSent = getSentMetadataFlag(sourcePort, sourceChannel, onChainDenom)
+      // if it hasn't been sent, then we will set the token metadata in the token
+      // so receiver can store it
+      // otherwise we leave metadata field blank to reduce redundancy
+      if !isMetadataSent {
+        metadata = getMetadata(onChainDenom)
+        token.metadata = metadata
+      }
+      if source {
+        // determine escrow account
+        escrowAccount = channelEscrowAddresses[sourceChannel]
+        // escrow source tokens (assumed to fail if balance insufficient)
+        bank.TransferCoins(sender, escrowAccount, onChainDenom, token.amount)
+      } else {
+        // receiver is source chain, burn vouchers
+        bank.BurnCoins(sender, onChainDenom, token.amount)
+      }
     }
 
     // create FungibleTokenPacket data
-    data = FungibleTokenPacketData{denomination, amount, sender, receiver}
+    data = FungibleTokenPacketData{tokens, sender, receiver, memo}
```

Comment:
Should there be here some kind of conditional logic based on the channel version?

```typescript
channel = provableStore.get(channelPath(sourcePort, sourceChannel))
if channel.version === "ics20-1" {
  data = FungibleTokenPacketData{tokens[0].denomination, tokens[0].amount, sender, receiver, memo}
} else if channel.version === "ics20-2" {
  data = FungibleTokenPacketDataV2{tokens, sender, receiver, memo}
}
```